Add plugins based CLI

hasbro17 opened the pull request:
**Description of the change:**
Add a new CLI entrypoint based on Kubebuilder plugins.
Currently uses plugins from the [feature/plugins-part-2-electric-boogaloo](https://github.com/kubernetes-sigs/kubebuilder/tree/feature/plugins-part-2-electric-boogaloo) branch until https://github.com/kubernetes-sigs/kubebuilder/pull/1469 is merged at which point we can update go.mod to use the master branch.

The `operator-sdk` CLI now has 2 entrypoints:
- The existing CLI (called `cmd/operator-sdk/cli/legacy.go` now) which has been updated with a hidden `operator-sdk init` command to allow scaffolding kubebuilder projects. See `operator-sdk init --project-version=2 -h`. 
- The Kubebuilder aligned CLI (`cmd/operator-sdk/cli/cli.go`) which only shows once you're inside a Kubebuilder project with a valid `PROJECT` file.

Once the integration is complete and we decide the CLI is ready for a release, we can change the order of the CLI entrypoints to make the Kubebulder aligned CLI the default, and deprecate the legacy CLI.



Also added `website/content/en/docs/kubebuilder/quickstart.md` as the new quickstart guide that is aligned with the new CLI and layout.
This quickstart guide defers to upstream docs as much as possible since the kubebuilder docs are detailed enough for further reading.
This doc folder will remain hidden from the website, since it has no index, until the integration and corresponding docs are completed.

Some of the advanced sections like metrics and kube-state CR metrics don't apply directly anymore since the Kubebuilder default scaffolding doesn't use SDK helpers for metrics. 
So I've marked those as TODOs and we'll need to figure out how to update/remove those docs.

estroz commented on cmd/operator-sdk/cli/cli.go:
```diff
@@ -42,91 +31,71 @@ import (
+	// New command is only present to allow scaffolding the legacy layout
+	// Once the legacy project is scaffolded, the legacy root CLI is used
+	// KB_INTEGRATION_TODO(estroz): remove this after 2 versions of deprecation
+	new.NewCmd(),
```
May want to set `new.NewCmd().Deprecated = "use 'operator-sdk init' instead"` if `Run()` is used.

hasbro17 commented on cmd/operator-sdk/main.go:
```diff
@@ -22,21 +22,25 @@ import (
+	// a Kubebuilder project as a way to opt into the new KB CLI.
+	// TODO: Make the new KB CLI the default, once the integration is complete
+	// and deprecate "operator-sdk new" from the old CLI.
+	if kbutil.IsConfigExist() {
```
@estroz This is what I meant by reversing the default CLI from what you had in #2779
With this, the legacy CLI is still the default when you're scaffolding a new project.
And the legacy CLI now provides the `operator-sdk init` command as a hidden subcommand after which you switch over to the new CLI.
https://github.com/operator-framework/operator-sdk/pull/2827/files#diff-48b64768aa84f740e1b924dffaf0f205R94

Let me know what you think.

estroz replied:
Ah I see you did the opposite of #2779. That's fine with me.

hasbro17 commented:
This can finally use some reviews.

In particular please go through the quickstart guide to verify the steps (skip the advanced sections), and look for typos/rephrasing etc.
If you have any suggestions for some of the TODOs I've marked feel free to comment although in most cases that's going to be addressed in follow up PRs to keep the scope of this PR limited to the new CLI workflow.


I am having some trouble with the sample-memcached example in the last step where the reconciler is failing to update the CR status:
```console
2020-04-14T21:14:42.978-0700    ERROR    controllers.Memcached    Failed to update Memcached status    {"memcached": "default/memcached-sample", "error": "memcacheds.cache.example.com \"memcached-sample\" not found"}
```
Debugging this to see if it isn't just my Kind cluster but I could use a verification on the updated sample-memcached reconciler code.

joelanford commented on cmd/operator-sdk/cli/cli.go:
```diff
@@ -42,91 +30,71 @@ import (
+	scorecard.NewCmd(),
+	test.NewCmd(),
+	version.NewCmd(),
+	// TODO(hasbro17): add generate csv command after aligning it for kubebuilder layout
```
> We could take advantage and create the testdata dir to scaffold a project to test it as it is done in the kubebuilder

IMO we shouldn't duplicate kubebuilder tests. Once we get to the point of making additional tweaks or changes to the Kubebuilder scaffolds or when we add our own plugins with custom layouts, it might make sense to add similar tests. Until then, I don't see how it adds any value, unless I'm misunderstanding something.

> Also, I do not think that we can merge this PR in the master until plugin phase 1 be merged there as well.

Typically I would agree with this, but I think in this case, it might make sense to push forward on this since:
* We're not making the KB CLI the default yet
* We're just now introducing the KB dependency, so our existing legacy code is not affected by the use of a feature branch commit.

What do others think?

camilamacedo86 replied:
HI @joelanford, 

I got your point of view over merge based on the branch and then it is OK for me. 
Regards the tests I really think that we should add them as soon as possible because as you said now we have nothing on top but we will have it. However, IHMO it is not a blocker to merge this PR at all. Would just make easier perform the review.   I am checking it manually anyway.

joelanford commented on website/content/en/docs/kubebuilder/quickstart.md:
````diff
@@ -0,0 +1,763 @@
+
+### Update the size
+
+Update `config/samples/cache_v1alpha1_memcached.yaml` to change the `spec.size` field in the Memcached CR from 3 to 4:
+
+```YAML
+apiVersion: cache.example.com/v1alpha1
+kind: Memcached
+metadata:
+  name: memcached-sample
+spec:
+  size: 3
+```
+
+Apply the change:
+
+```
+$ kubectl apply -f config/samples/cache_v1alpha1_memcached.yaml
+```
````
Simplify this to use `kubectl patch`?

```
kubectl patch memcached memcached-example -p '{"spec":{"size": 4}}'
```

joelanford replied:
Not positive that this works actually. Might need this instead:
```
kubectl patch memcached memcached-sample -p '[{"op":"replace","path":"/spec/size","value":4}]' --type=json

hasbro17 replied:
This works too:
```
$ kubectl patch memcached memcached-sample -p '{"spec":{"size": 4}}' --type="merge"
```
Since we're just replacing the field we can specify `--type=merge` which is a [JSON merge patch](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#use-a-json-merge-patch-to-update-a-deployment).

joelanford commented on website/content/en/docs/kubebuilder/quickstart.md:
````diff
@@ -0,0 +1,763 @@
+$ kubectl delete role,rolebinding --all
+```
+
+## Advanced Topics
````
Move this section to a separate page and perhaps link to it here?

joelanford commented on website/content/en/docs/kubebuilder/quickstart.md:
```diff
@@ -0,0 +1,763 @@
+linkTitle: Quickstart
+weight: 1
+---
+
```
For now, should we include a prominent note at the top saying something about this being an alpha feature, and maybe a link over to the main quickstart?

jmccormick2001 commented:
Foo string `json:"foo,omitempty"`

in testing this, I noticed that the above get generated by default...expected?  I guess it doesn't matter but users might ask about it.

jmccormick2001 commented:
For this example replace the generated controller file controllers/memcached_controller.go with the example memcached_controller.go implementation.

the link to memcached_controller.go appears to be broken in the quickstart doc.

jmccormick2001 commented:
the quickstart guide doesn't specifically call out that you need to install kustomize, that might be good at the top of the quickstart unless I missed it somewhere else.

jmccormick2001 commented:
For this example we will run the operator in the default namespace which can be specified for all resources in config/default/kustomization.yaml:

# Adds namespace to all resources.
namespace: default

in testing, I noticed that it installs the operator into the namespace of 'memcached-operator-system' instead of default, so the quick start guide might need to be updated.

camilamacedo86 commented on cmd/operator-sdk/cli/cli.go:
```diff
@@ -42,91 +30,71 @@ import (
 	"github.com/spf13/viper"
+	"sigs.k8s.io/kubebuilder/pkg/cli"
+	kbgov1 "sigs.k8s.io/kubebuilder/pkg/plugin/v1"
+	kbgov2 "sigs.k8s.io/kubebuilder/pkg/plugin/v2"
```
Hi @estroz, @hasbro17 

The Project file scaffolded from here is: 

```
$ cat PROJECT 
domain: my.domain
layout: go.kubebuilder.io/v2.0.0
repo: test
version: 3-alpha

``` 

-  Should not the layout be `go.operator-sdk.io/v3-alpha`?

estroz replied:
We're using kubebuilder's go plugin, which sets the layout.

Note: plugin version (`v2.0.0`) is different from project version (`v3-alpha`).

camilamacedo86 replied:
How we will know if the project was scaffolded by sdk or kubebuilder tool? 
I thought that this info would be in the layout.

estroz replied:
Since we're eventually going to want Makefile targets for generating and pushing bundles, deploying the operator etc, we may have to wrap the base kubebuilder plugin in our own. Then `layout` will be something like `go.operator-sdk.io/v2.0.0`. Out of scope for this PR though.

camilamacedo86 replied:
Solved

camilamacedo86 commented on cmd/operator-sdk/cli/legacy.go:
```diff
@@ -0,0 +1,160 @@
+	// The "init" cmd is is added as a hidden subcommand from the new KB CLI to provide
+	// a way to initialize Kubebuilder projects from the legacy CLI.
+	// TODO(hasbro17): This should be removed once the KB CLI is the default.
+	initCmd := getInitCmd()
```
Init with the domain is not working. 

```
$ operator-sdk init --domain my.domain
Error: failed to initialize project with version "3-alpha": error finding current repository: could not determine repository path from module data, package data, or by initializing a module: go: cannot determine module path for source directory /Users/camilamacedo/go/src (outside GOPATH, module path must be specified)

Example usage:
	'go mod init example.com/m' to initialize a v0 or v1 module
	'go mod init example.com/m/v2' to initialize a v2 module

Run 'go help mod init' for more information.

Usage:
  operator-sdk init [flags]

Examples:
  # Scaffold a project using the apache2 license with "The Kubernetes authors" as owners
  operator-sdk init --project-version=2 --domain example.org --license apache2 --owner "The Kubernetes authors"


Flags:
      --domain string            domain for groups (default "my.domain")
      --fetch-deps               ensure dependencies are downloaded (default true)
  -h, --help                     help for init
      --license string           license to use to boilerplate, may be one of 'apache2', 'none' (default "apache2")
      --owner string             owner to add to the copyright
      --project-version string   project version, possible values: ("2", "3-alpha") (default "3-alpha")
      --repo string              name to use for go module (e.g., github.com/user/repo), defaults to the go package of the current working directory.
      --skip-go-version-check    if specified, skip checking the Go version

Global Flags:
      --verbose   Enable verbose logging
```

c/c @estroz @hasbro17

estroz replied:
Where are you running your command and what go version are you using?

hasbro17 replied:
@camilamacedo86 Did you run `go mode init` before `operator-sdk init`?
Or use the `--repo` flag?

camilamacedo86 replied:
Let's close this one .. I am unable to see the error again. I probably did something wrong.

camilamacedo86 commented on cmd/operator-sdk/cli/cli.go:
```diff
@@ -42,91 +30,71 @@ import (
+	cleanup.NewCmd(),
+	completion.NewCmd(),
+	olm.NewCmd(),
+	run.NewCmd(),
```
HI @estroz and @hasbro17, 
  
The go.mod scaffolded should not have the sdk modules? 
See its result:

```
 $ cat go.mod 
module example.com/m/v2

go 1.13

require (
	github.com/go-logr/logr v0.1.0
	github.com/onsi/ginkgo v1.11.0
	github.com/onsi/gomega v1.8.1
	k8s.io/apimachinery v0.17.2
	k8s.io/client-go v0.17.2
	sigs.k8s.io/controller-runtime v0.5.0
)


```

estroz replied:
The base go scaffold in kubebuilder does not use operator-sdk packages.

hasbro17 replied:
@camilamacedo86 Right now nothing in the default scaffold is importing anything from the SDK libraries (e.g `pkg/log/zap` or `pkg/metrics`) so the operator-sdk wouldn't really show up in go.mod. Much like `sigs.k8s.io/kubebuilder` doesn't show up in kubebuilder projects.

This may change if we decide to override the default plugin scaffolds to change anything later on.

camilamacedo86 replied:
I think would be nice we add our scaffolds and etc as add the sdk modules to allow the users to use it. But it is out of the scope of this PR.. so no worry. tks for the clarifications.

estroz commented on internal/util/kubebuilder/project.go:
```diff
@@ -0,0 +1,65 @@
+	}
+}
+
+func ReadConfig() (*config.Config, error) {
```
@hasbro17 this can be removed since the `CLI` handles config reading/writing. It can be added back later if needed.

hasbro17 replied:
Oh yeah that's a leftover from your other PR. Good catch.

estroz commented on cmd/operator-sdk/cli/cli.go:
```diff
@@ -42,91 +30,71 @@ import (
+	c, err := cli.New(
+		cli.WithCommandName("operator-sdk"),
+		cli.WithPlugins(
+			&kbgov1.Plugin{},
```
We probably don't want to support kubebuilder v1 projects..

camilamacedo86 replied:
Yep. It is deprecated for too long now and the code will be removed. 
We will probably not have the v1.Plugin.

hasbro17 commented:
@jmccormick2001 Addressing your comments here but for follow ups it'll probably be better to add direct review comments on the PR files so I can respond to them more easily.

> Foo string `json:"foo,omitempty"`
> in testing this, I noticed that the above get generated by default...expected?

That's the default scaffold in Kubebuilder for the sample CR so it's expected.

> the link to memcached_controller.go appears to be broken in the quickstart doc.

That link is currently pointing to the file on the master branch which should work once this PR is merged.
https://github.com/operator-framework/operator-sdk/blob/master/example/kb-memcached-operator/memcached_controller.go.tmpl

> the quickstart guide doesn't specifically call out that you need to install kustomize

Yep good point. I'll add that to the pre-reqs.

> in testing, I noticed that it installs the operator into the namespace of 'memcached-operator-system' instead of default, so the quick start guide might need to be updated.

I'll make the following statement more actionable to get users to actually change the kustomization.yaml file so they can run in the default namespace. Probably with `kustomize edit` if possible.
```
For this example we will run the operator in the default namespace which can be specified for all resources in config/default/kustomization.yaml
```
```YAML
# Adds namespace to all resources.
namespace: default
```

estroz commented on website/content/en/docs/kubebuilder/quickstart.md:
````diff
@@ -0,0 +1,763 @@
+
+```sh
+$ kubectl delete -f config/samples/cache_v1alpha1_memcached.yaml
+$ $ kubectl delete deployments,service -l control-plane=controller-manager
````
```suggestion
$ kubectl delete deployments,service -l control-plane=controller-manager
```

estroz commented on website/content/en/docs/kubebuilder/quickstart.md:
````diff
@@ -0,0 +1,763 @@
+
+Verify that the memcached-operator is up and running:
+
+```sh
````
Anything run from the shell with output looks better with a `console` tag:
```console
$ kubectl get deployment
NAME                                    READY   UP-TO-DATE   AVAILABLE   AGE
memcached-operator-controller-manager   1/1     1            1           8m
```
vs
```sh
$ kubectl get deployment
NAME                                    READY   UP-TO-DATE   AVAILABLE   AGE
memcached-operator-controller-manager   1/1     1            1           8m
```

estroz commented on cmd/operator-sdk/cli/legacy.go:
```diff
@@ -0,0 +1,160 @@
+	return root
+}
+
+func getInitCmd() *cobra.Command {
```
Non-blocking note: the way `init` is retrieved prevents global flags set by root, which contain `--plugins`, from being retrieved. This is not blocking because right now only one plugin is available. Once https://github.com/kubernetes-sigs/kubebuilder/pull/1465 merges, `--plugins` will be local to `init`, so we can bump kubebuilder commits to get `--plugins` registration.

hasbro17 commented on website/content/en/docs/kubebuilder/quickstart.md:
````diff
@@ -0,0 +1,763 @@
+
+Build the image:
+
+```sh
+$ make docker-build IMG=quay.io/example/memcached-operator:v0.0.1
+```
+
+Push the image to a repository:
+
+```sh
+$ make docker-push IMG=quay.io/example/memcached-operator:v0.0.1
+```
````
Yeah this is better. 
Makes it so that they can't blindly copy paste to build but then be unable to push the image.

I'll still add a line about replacing the image line completely depending on the image repository that they're using.

hasbro17 commented:
Updated for review comments. Had to rebase for a conflict from master though.

estroz commented on website/content/en/docs/kubebuilder/quickstart.md:
```diff
@@ -0,0 +1,457 @@
+$ cd $HOME/projects/memcached-operator
+# we'll use a domain of example.com
+# so all API groups will be <group>.example.com
+$ operator-sdk init memcached-operator --domain=example.com --repo=github.com/example-inc/memcached-operator
```
```suggestion
$ operator-sdk init --domain=example.com --repo=github.com/example-inc/memcached-operator
```
